From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Anthony Ramine <nox@nox.paris>
Date: Tue, 7 Jan 2020 12:34:28 +0100
Subject: [PATCH] Don't include C define flags in assembler arguments


diff --git a/python/mozbuild/mozbuild/frontend/emitter.py b/python/mozbuild/mozbuild/frontend/emitter.py
index 1d73d0574a64..aa5233249b08 100644
--- a/python/mozbuild/mozbuild/frontend/emitter.py
+++ b/python/mozbuild/mozbuild/frontend/emitter.py
@@ -1186,7 +1186,7 @@ class TreeMetadataEmitter(LoggingMixin):
             yield sub
 
         for defines_var, cls, backend_flags in (
-                ('DEFINES', Defines, (computed_flags, computed_as_flags,)),
+                ('DEFINES', Defines, (computed_flags,)),
                 ('HOST_DEFINES', HostDefines, (computed_host_flags,)),
                 ('WASM_DEFINES', WasmDefines, (computed_wasm_flags,))):
             defines = context.get(defines_var)
