From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Anthony Ramine <nox@nox.paris>
Date: Fri, 24 Jan 2020 16:47:59 +0100
Subject: [PATCH] Disable spidermonkey style checks

This fails because we removed cranelift from the tree.

diff --git a/js/src/build/moz.build b/js/src/build/moz.build
index b54730ca28bc..5c1c8122f8a4 100644
--- a/js/src/build/moz.build
+++ b/js/src/build/moz.build
@@ -89,7 +89,6 @@ GeneratedFile(
     py2=True,
     inputs=[
         '!%sjs_static.%s' % (CONFIG['LIB_PREFIX'], CONFIG['LIB_SUFFIX']),
-        '/config/check_spidermonkey_style.py',
         '/config/check_macroassembler_style.py',
         '/config/check_js_opcode.py'
         ])
